lib/service/awsoidc.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	if cfg.Log == nil {
+		cfg.Log = logrus.WithField(trace.Component, teleport.Component(teleport.ComponentProxy, "aws_oidc_deploy_service_updater"))
+	}
+
+	if cfg.Clock == nil {
+		cfg.Clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+// DeployServiceUpdater periodically updates deploy service agents
+type DeployServiceUpdater struct {
+	DeployServiceUpdaterConfig
+}
+
+// NewDeployServiceUpdater returns a new DeployServiceUpdater
+func NewDeployServiceUpdater(config DeployServiceUpdaterConfig) (*DeployServiceUpdater, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &DeployServiceUpdater{
+		DeployServiceUpdaterConfig: config,
+	}, nil
+}
+
+// Run periodically updates the deploy service agents
+func (updater *DeployServiceUpdater) Run(ctx context.Context) error {
+	periodic := interval.New(interval.Config{
+		Duration: updateDeployAgentsInterval,
+		Jitter:   retryutils.NewSeventhJitter(),
+	})
+	defer periodic.Stop()
+
+	for {
+		if err := updater.updateDeployServiceAgents(ctx); err != nil {
+			updater.Log.WithError(err).Warningf("Update failed. Retrying in ~%v.", updateDeployAgentsInterval)
+		}
+
+		select {
+		case <-periodic.Next():
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (updater *DeployServiceUpdater) updateDeployServiceAgents(ctx context.Context) error {
+	cmc, err := updater.AuthClient.GetClusterMaintenanceConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	critical, err := automaticupgrades.Critical(ctx, updater.CriticalEndpoint)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Upgrade should only be attempted if the current time is within the configured
+	// upgrade window, or if a critical upgrade is available
+	if !cmc.WithinUpgradeWindow(updater.Clock.Now()) && !critical {
+		return nil
+	}
+
+	stableVersion, err := automaticupgrades.Version(ctx, updater.VersionEndpoint)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// stableVersion has vX.Y.Z format, however the container image tag does not include the `v`.
+	stableVersion = strings.TrimPrefix(stableVersion, "v")
+
+	databases, err := updater.AuthClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The updater needs to iterate over all integrations and aws regions to check
+	// for deploy service agents to update. In order to reduce the number of api
+	// calls, the aws regions are first reduced to only the regions containing
+	// an RDS database.
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	integrations, err := updater.AuthClient.ListAllIntegrations(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Perform updates in parallel across regions.
+	var sem = make(chan interface{}, maxConcurrentUpdates)
+	var wg sync.WaitGroup
```

Comment:
:nail_care: 
Maybe we should use
https://pkg.go.dev/golang.org/x/sync/semaphore#example-package-WorkerPool instead

Sorry for recommending a WaitGroup